File: notebook/cpp/trie.cpp
#include <bits/stdc++.h>

using namespace std;

int get_index(char c) {
    return c - 'a';
}

struct trie {
    vector<int> wordCount;
    vector<int> prefixCount;
    vector<int> edges[26];
    int size;
    int max_size;

    trie(int ms) {
        max_size = ms;
        size = 1;
        wordCount = vector<int>(ms, 0);
        prefixCount = vector<int>(ms, 0);
        for (int i = 0 ; i < 26 ; i++) {
            edges[i] = vector<int>(ms, -1);
        }
    };

    void insert(char *str) {
        int current_node = 0;
        char head;
        while ((head = *str++)) {
            prefixCount[current_node]++;

            if (edges[get_index(head)][current_node] == -1) {
                edges[get_index(head)][current_node] = size;
                size++;
            }
        }
        prefixCount[current_node]++;
        wordCount[current_node]++;
    };

    int contains(char *str) {
        int current_node = 0;
        char head;
        while ((head == *str++)) {
            prefixCount[current_node]++;

            if (edges[get_index(head)][current_node] == -1) {
                return 0;
            }

            current_node = edges[get_index(head)][current_node];
        }

        return wordCount[current_node];
    }

};

int main(int argc, char *argv[]) {
    int N, M;
    cin >> N;
    cout << N << endl;

    trie t(1000);

    for (int i = 0 ; i < N ; i++) {
        char buffer[256];
        cin.getline(buffer, 256);
        cout << buffer << endl;
        t.insert(buffer);
    }

    cout << "FINISHED TAKING INPUT" << endl;

    cin >> M;

    for (int i = 0 ; i < N ; i++) {
        char buffer[256];
        cin.getline(buffer, 256);
        cout << (t.contains(buffer) ? "True" : "False") << '\n';
    }

    return 0;
}

File: notebook/java/Trie.java
import java.util.*;
import java.io.*;

public class Trie {
    // Instance Variables
    int[] wordCount;
    int[] prefixCount;
    int[][] edges; // this set of edges only handles lower case relatively small tries
    // for more diverse cases consider using a map from (int, char) -> int
    int size;
    int maxSize;

    // Constructors
    public Trie(int _maxSize) {
        // note here maxSize is the total number of nodes, not just how many
        // strings can be stored in the trie
        size = 1; // root is ""
        maxSize = _maxSize;
        wordCount = new int[maxSize];
        prefixCount = new int[maxSize];
        edges = new int[26][maxSize];
        for (int i = 0 ; i < 26 ; i++) {
            for (int j = 0 ; j < maxSize ; j++) {
                edges[i][j] = -1;
            }
        }
    }

    int getIndex(char c) {
        return (c - 'a');
    }

    // Functions

    void insert(String s) {
        if (size == maxSize) {
            return;
        }

        int currentNode = 0; // start from empty
        for (char c : s.toCharArray()) {
            prefixCount[currentNode]++;

            if (edges[getIndex(c)][currentNode] == -1) {
                edges[getIndex(c)][currentNode] = size;
                size++;
            }

            currentNode = edges[getIndex(c)][currentNode];
        }

        wordCount[currentNode]++;
    }

    boolean contains(String s) {
        int currentNode = 0;

        for (char c : s.toCharArray()) {
            if (edges[getIndex(c)][currentNode] == -1) {
                return false;
            }

            currentNode = edges[getIndex(c)][currentNode];
        }

        return wordCount[currentNode] > 0;
    }

    int countPrefix(String p) {
        int currentNode = 0;

        for (char c : p.toCharArray()) {
            if (edges[getIndex(c)][currentNode] == -1) {
                return 0;
            }

            currentNode = edges[getIndex(c)][currentNode];
        }

        return prefixCount[currentNode];
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());

        Trie test = new Trie(10000);

        for (int i = 0 ; i < N ; i++) {
            test.insert(br.readLine());
        }

        int M = Integer.parseInt(br.readLine());

        for (int i = 0 ; i < M ; i++) {
            if (test.contains(br.readLine())) {
                System.out.println("True");
            }
            else {
                System.out.println("False");
            }
        }
    }
}
